### PR TITLE
REGR: fix op(frame, frame2) with reindex

### LINKED ISSUE #31623: REGR: AssertionError when subtracting Timestamp-valued DataFrames with non-indentical column index
```python
import pandas as pd

df = pd.DataFrame(
    {
        "foo": [pd.Timestamp("2019"), pd.Timestamp("2020")],
        "bar": [pd.Timestamp("2018"), pd.Timestamp("2021")],
    }
)

df2 = df[["foo"]]

print(df - df2)
```
#### Problem description

The above snippet raises the following exception:

```
Traceback (most recent call last):
  File ".venv/lib/python3.6/site-packages/pandas/core/ops/array_ops.py", line 149, in na_arithmetic_op
    result = expressions.evaluate(op, str_rep, left, right)
  File ".v
env/lib/python3.6/site-packages/pandas/core/computation/expressions.py", line 208, in evaluate
    return _evaluate(op, op_str, a, b)
  File ".venv/lib/python3.6/site-packages/pandas/core/computation/expressions.py", line 70, in _evaluate_standard
    return op(a, b)
  File ".venv/lib/python3.6/site-packages/pandas/core/ops/common.py", line 64, in new_method
    return method(self, other)
  File ".venv/lib/python3.6/site-packages/pandas/core/ops/__init__.py", line 500, in wrapper
    result = arithmetic_op(lvalues, rvalues, op, str_rep)
  File ".venv/lib/python3.6/site-packages/pandas/core/ops/array_ops.py", line 192, in arithmetic_op
    res_values = dispatch_to_extension_op(op, lvalues, rvalues)
  File ".venv/lib/python3.6/site-packages/pandas/core/ops/dispatch.py", line 125, in dispatch_to_extension_op
    res_values = op(left, right)
  File ".venv/lib/python3.6/site-packages/pandas/core/arrays/datetimelike.py", line 1390, in __rsub__
    f"cannot subtract {type(self).__name__} from {type(other).__name__}"
TypeError: cannot subtract DatetimeArray from ndarray

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "pandas_bug.py", line 36, in <module>
    print(df2 - df)
  File ".venv/lib/python3.6/site-packages/pandas/core/ops/__init__.py", line 703, in f
    new_data = left._combine_frame(right, pass_op, fill_value)
  File ".venv/lib/python3.6/site-packages/pandas/core/frame.py", line 5297, in _combine_frame
    new_data = ops.dispatch_to_series(self, other, _arith_op)
  File ".venv/lib/python3.6/site-packages/pandas/core/ops/__init__.py", line 416, in dispatch_to_series
    new_data = expressions.evaluate(column_op, str_rep, left, right)
  File ".venv/lib/python3.6/site-packages/pandas/core/computation/expressions.py", line 208, in evaluate
    return _evaluate(op, op_str, a, b)
  File ".venv/lib/python3.6/site-packages/pandas/core/computation/expressions.py", line 70, in _evaluate_standard
    return op(a, b)
  File ".venv/lib/python3.6/site-packages/pandas/core/ops/__init__.py", line 385, in column_op
    return {i: func(a.iloc[:, i], b.iloc[:, i]) for i in range(len(a.columns))}
  File ".venv/lib/python3.6/site-packages/pandas/core/ops/__init__.py", line 385, in <dictcomp>
    return {i: func(a.iloc[:, i], b.iloc[:, i]) for i in range(len(a.columns))}
  File ".venv/lib/python3.6/site-packages/pandas/core/ops/array_ops.py", line 121, in na_op
    return na_arithmetic_op(x, y, op, str_rep)
  File ".venv/lib/python3.6/site-packages/pandas/core/ops/array_ops.py", line 151, in na_arithmetic_op
    result = masked_arith_op(left, right, op)
  File ".venv/lib/python3.6/site-packages/pandas/core/ops/array_ops.py", line 75, in masked_arith_op
    assert isinstance(x, np.ndarray), type(x)
```
This is a 1.0.0 regression; in 0.25.3, the operation succeeds and the unmatched `bar` column is filled with `NaN` in the output.

The same error occurs with:
* Any combination of incompatible columns (strict subset, strict superset, overlapping, disjoint)
* Calling the `subtract` method instead of using the subtraction operator
* Timezone-aware `Timestamp`s as well as timezone-naive

It does *not* seem to occur with:
* Mismatches on the row index; transposing the dataframes in the above example prevents the errors occuring.
* `pd.Series` objects with mismatched indexes (e.g. calling the above on the first row of each dataframe works fine)
* Other dtypes; `bool`, `float`, and `int` seem to work fine. Similarly, if the dataframes are explicitly cast to dtype `object`, the operation succeeds.

#### Expected Output

```
   bar    foo
0  NaN 0 days
1  NaN 0 days
```

#### Output of ``pd.show_versions()``

<details>
```
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.6.8.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.15.0-74-generic
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.0.0
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 41.6.0
Cython           : None
pytest           : 5.3.5
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : 5.3.5
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None
```

</details>

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #31623
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

cc @TomAugspurger this is pretty ugly, and I'm not sure how well it will behave if either frame has MultiIndex colums.

On the plus side, it could improve perf in the many-columns-but-small-intersection case.

The ugliness might be improved by moving this check to before the _align_method_FRAME call

### CODE REVIEW

File: pandas/core/ops/__init__.py [hunk 1]
```diff
@@ -722,13 +722,28 @@ def _arith_method_FRAME(cls, op, special):
...
+                    cols = orig_self.columns.intersection(orig_other.columns)
+                    if not (
+                        cols.equals(orig_self.columns)
+                        and cols.equals(orig_other.columns)
+                    ):
+                        # GH#31623, only operate on shared columns
+                        new_left = orig_self[cols]
+                        new_right = orig_other[cols]
+                        result = op(new_left, new_right)
+                        return result.reindex(self.columns, axis=1)
```
Comment (TomAugspurger) [inline pandas/core/ops/__init__.py]: This should be the union of the columns right?

Comment (jbrockmendel) [reply pandas/core/ops/__init__.py]: i think so, but id have to double-check the exact outer-join semantics

**Comment by jorisvandenbossche** [comment]
@jbrockmendel is this still a POC? Or is this good for 1.0.2?

**Comment by jbrockmendel** [comment]
> is this still a POC? Or is this good for 1.0.2?

Borderline.  Do we have a timeline for 1.0.2?

**Comment by TomAugspurger** [comment]
Next week-ish?

**Comment by jbrockmendel** [comment]
> Next week-ish?

Sounds good.  I'd like to try a second draft of this, if we're OK with this approach.

**Comment by jorisvandenbossche** [comment]
The approach looks generally OK to me. 
But for possible alternative approach: do you know why this started failing? What changed in the ops code that might have caused this?

File: pandas/core/ops/__init__.py [hunk 2]
```diff
@@ -722,13 +722,28 @@ def _arith_method_FRAME(cls, op, special):
...
 
+        orig_self, orig_other = self, other
         self, other = _align_method_FRAME(self, other, axis, flex=True, level=level)
 
         if isinstance(other, ABCDataFrame):
             # Another DataFrame
             pass_op = op if should_series_dispatch(self, other, op) else na_op
             pass_op = pass_op if not is_logical else op
 
+            if isinstance(orig_other, ABCDataFrame):
```
Comment (jreback) [inline pandas/core/ops/__init__.py]: shouldn't this be handled in .align? why is this needed at all?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jorisvandenbossche** [comment]
@jbrockmendel What's the status of this?

**Comment by TomAugspurger** [comment]
I opened https://github.com/pandas-dev/pandas/issues/31874 for a larger discussion on how to fix this, which addresses Jeff's comment in https://github.com/pandas-dev/pandas/pull/31679/files#r376807594. That's too large to do now though.

**Comment by jbrockmendel** [comment]
> What's the status of this?

I'm catching up on reviewing things, hope to push a second draft of this later today.  I don't expect the logic to change, just the organization.

**Comment by jbrockmendel** [comment]
Updated, separating out the new logic to dedicated functions, doing it before the _align_frame_METHOD call.

File: pandas/core/ops/__init__.py [hunk 3]
```diff
@@ -705,6 +705,52 @@ def to_series(right):
     return left, right
 
 
+def _should_reindex_op(left, right, axis, default_axis, fill_value, level) -> bool:
+    """
+    Check if this is an operation between DataFrames that will need to reindex.
+    """
+    if not isinstance(right, ABCDataFrame):
```
Comment (jreback) [inline pandas/core/ops/__init__.py]: what about left? (can just assert that let is a DF)

File: pandas/core/ops/__init__.py [hunk 4]
```diff
@@ -705,6 +705,52 @@ def to_series(right):
...
+        return False
+
+    if fill_value is None and level is None and axis is default_axis:
+        # TODO: any other cases we should handle here?
+        cols = left.columns.intersection(right.columns)
```
Comment (jreback) [inline pandas/core/ops/__init__.py]: isn't this just
set(left.columns) == set(right.columns) ?

File: pandas/core/ops/__init__.py [hunk 5]
```diff
@@ -705,6 +705,52 @@ def to_series(right):
     return left, right
 
 
+def _should_reindex_op(left, right, axis, default_axis, fill_value, level) -> bool:
```
Comment (jreback) [inline pandas/core/ops/__init__.py]: maybe add _frame_ in the name here

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
looks good, some comments

**Review by jorisvandenbossche — COMMENTED** [review]
Can you add a whatsnew?

File: doc/source/whatsnew/v1.0.2.rst [hunk 6]
```diff
@@ -19,6 +19,7 @@ Fixed regressions
 - Fixed regression in :meth:`Series.align` when ``other`` is a DataFrame and ``method`` is not None (:issue:`31785`)
 - Fixed regression in :meth:`pandas.core.groupby.RollingGroupby.apply` where the ``raw`` parameter was ignored (:issue:`31754`)
 - Fixed regression in :meth:`rolling(..).corr() <pandas.core.window.Rolling.corr>` when using a time offset (:issue:`31789`)
+- Fixed regression in :class:`DataFrame` arithmetic operations with mis-matched columns (:issue:`31623`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.2.rst]: does this only affect datetime based ones, or generally?

File: pandas/core/ops/__init__.py [hunk 7]
```diff
@@ -703,6 +703,56 @@ def to_series(right):
     return left, right
 
 
+def _should_reindex_frame_op(
+    left, right, axis, default_axis, fill_value, level
```
Comment (jreback) [inline pandas/core/ops/__init__.py]: try to type when you can

File: pandas/core/ops/__init__.py (hunk 4 above)
Comment (jreback) [reply pandas/core/ops/__init__.py]: ?

**Comment by jreback** [comment]
@jbrockmendel couple of comments.

**Comment by jbrockmendel** [comment]
I'm under the weather, will get to these once I'm back on my feet.

**Comment by jreback** [comment]
> I'm under the weather, will get to these once I'm back on my feet.

sure np! feel better!

File: doc/source/whatsnew/v1.0.2.rst (hunk 6 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v1.0.2.rst]: the only example im aware of was for datetime col

File: pandas/core/ops/__init__.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/ops/__init__.py]: yah, but for certain types of indexes (e.g. RangeIndex) intersection is optimized.  Usually won't be enough to matter, but still

**Comment by jbrockmendel** [comment]
comments addressed i think

File: pandas/core/ops/__init__.py (hunk 4 above)
Comment (TomAugspurger) [reply pandas/core/ops/__init__.py]: Do you know if `xor` also optimized? I think this is equivalent to `not len(left.columns ^ right.columns)`. My guess is that xor will tend to be slower than your check.

File: pandas/core/ops/__init__.py [hunk 8]
```diff
@@ -703,6 +706,58 @@ def to_series(right):
...
+    """
+    # GH#31623, only operate on shared columns
+    cols = left.columns.intersection(right.columns)
+
+    new_left = left[cols]
+    new_right = right[cols]
+    result = op(new_left, new_right)
+
+    # Do the join on the columns instead of using _align_method_FRAME
+    #  to avoid constructing two potentially large/sparse DataFrames
```
Comment (TomAugspurger) [inline pandas/core/ops/__init__.py]: This is a nice side-effect of this bugfix.

**Review by TomAugspurger — APPROVED** [review]
I think we're good here.

File: pandas/core/ops/__init__.py (hunk 4 above)
Comment (jreback) [reply pandas/core/ops/__init__.py]: ok, fair point. yeah we likley just need some more asv's around this (followup)

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**